### DISCUSSION TITLE
Livewire 3 FormObjects with Livewire-Wizard

### DISCUSSION THREAD

**Discussion by jurnskie** [spatie/laravel-livewire-wizard#100] [Q&A]
Hi ! 

I'm trying to use [FormObjects](https://livewire.laravel.com/docs/forms) with Livewire Wizard. At first ik looks like its working great. But when navigating back to a previous step its trying to assign the FormObject as array onto the typed Form class of Livewire.

Steps to reproduce:

- Create formobject and use it in a step component
- Navigate to next step
- Go to previous step

You will be presented with the following error:

`Cannot assign array to property App\Livewire\AppointmentStepComponent\Steps\UserDetailsStep::$form of type App\Livewire\Forms\UserDetailsForm
`
Anyone else run into this?

**Comment by Hassan0x00A** [comment]
Hi. I was doing just that in my project and got the same error. Did you find any solutions?

**Reply by jurnskie** [reply]
I did not find any solutions as of yet.

**Comment by GregOstry** [comment]
I'm getting same error when I attempt to navigate to the previous step.

**Reply by jurnskie** [reply]
yeah i think the form object needs to be transformed  to an array, but i dont know how yet.

**Comment by jurnskie** [comment]
I started trying again and resolved the error by defining the form like this, i dont know yet if the data is persisted though

```
 public $form;

    public function mount()
    {
        $this->form =  New UserForm($this, 'form');
    }
```

**Comment by kodyxgen** [comment]
Hi @jurnskie did you find a solution for the problem ? I am getting the same error :(

**Reply by jurnskie** [reply]
Hey! Did you try this?

```
 public $form;

    public function mount()
    {
        $this->form =  New UserForm($this, 'form');
    }
```

**Reply by jurnskie** [reply]
I thinks its a step in the right direction in terms of finding the answer.

**Reply by kodyxgen** [reply]
Hi, 
yes i was playing with that, and got it to work like this, but i don't know if this is the best way :)
```php
       $this->dataForm = new DataForm($this, 'dataForm');

        $data = $this->allStepsState;

        if (isset($data[$this->state()->getStepName()])) {
            $this->dataForm->fill($data[$this->state()->getStepName()]['dataForm']);
        }
```

**Reply by jurnskie** [reply]
if it works, then for now this is the best way :D

**Reply by kodyxgen** [reply]
Yes, it works! Thank you! 😁

**Comment by davitrix** [comment]
Muchas gracias, pero ninguna de las soluciones propuestas funciono para mí de echo la función "getStepName" parece no existir, quedo a la ansiosa espera.

**Comment by calvinkihoro** [comment]
That's true @davitrix. You can get a current step name using `php app(ComponentRegistry::class)->getName(static::class)`. 

Here is a simple trait i have created based on @kodyxgen solution.

```php
trait HasObjectForm
{
    public function mountWithObjectForm(string $formClass = null, string $formProperty = 'form')
    {
        if (!$formClass) {
            return;
        }
        $this->$formProperty = new $formClass($this, $formProperty);
        
        $states = $this->allStepsState;

        $currentComponentName = app(ComponentRegistry::class)->getName(static::class);

        if (isset($states[$currentComponentName])) {
            $this->$formProperty->fill($states[$currentComponentName][$formProperty]);
        }
    }
}
```

Then in your component file you can use it like 

```php

class UserDetailsStep extends StepComponent
{

    use HasObjectForm;

    public $form;

    public function mount()
    {
        $this->mountWithObjectForm(UserDetailsForm::class);
      
       //  $this->mountWithObjectForm(UserDetailsForm::class, 'form'); Or passing the form property name 

    }
```

**Reply by jurnskie** [reply]
Nice thanks for this :)
